### PR TITLE
Add "only" violation types options

### PR DESCRIPTION
**Pull request by oboxodo**
Adds options to allow todos by type. This is useful to generate diagrams focused on specific todos. In my case, I wanted to be able to generate a diagram showing only architecture layers' todos.

image

### CODE REVIEW

File: bin/visualize_packs [hunk 1]
```diff
@@ -26,6 +26,8 @@ OptionParser.new do |opt|
   opt.on('--no_nested_relationships', "Don't draw nested package relationships") { |o| options.show_nested_relationships = false }
 
   opt.on('--exclude-packs=pack1,pack2,etc', "Exclude these packs from diagram") { |o| options.exclude_packs = o.to_s.split(",") }
+  opt.on('--exclude-violation-types=privacy,architecture,etc', "Exclude these types of violations") { |o| options.exclude_violation_types = o.to_s.split(",") }
+  opt.on('--only-violation-types=privacy,architecture,etc', "Show only these types of violations") { |o| options.only_violation_types = o.to_s.split(",") }
```
[the remark this answers is not in the document]
Comment (shageman) [reply bin/visualize_packs:30]: Also, given that there are only a handful of todo types... could we get away with doing only one of these options?

[the remark this answers is not in the document]
Comment (shageman) [reply bin/visualize_packs:30]: The language around packs has moved to calling not calling them violations but rather "todos." This code should adopt that language.

[the remark this answers is not in the document]
Comment (shageman) [reply bin/visualize_packs:30]: I also believe these selectors should be grouped with the other todo option, not with the packs exclusion

File: lib/visualize_packs.rb [hunk 2]
```diff
@@ -76,14 +78,14 @@ def self.diagram_title(options, max_violation_count)
...
+    case list.size
     when 1
-      exclude_packs.first
+      list.first
     when 2
-      exclude_packs.join(" and ")
+      list.join(" and ")
     else
-      "#{exclude_packs[0, 2].join(", ")}, and #{exclude_packs.size - 2} more."
+      "#{list[0, 2].join(", ")}, and #{list.size - 2} more"
```
Comment (shageman) [inline lib/visualize_packs.rb:88]: I didn't catch this before, but a unit test for this method would be nice

**Review by shageman — COMMENTED** [review]
Nice addition! I have some feedback that I would like to discuss

File: bin/visualize_packs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oboxodo) [reply bin/visualize_packs:30]: @shageman:
> How does a user know which options exist? I think that should be added here.

I'll list the ones currently supported.

> Also, given that there are only a handful of todo types... could we get away with doing only one of these options?

I think it's worth supporting both allow and deny lists. But I understand the POV on only adding features as they're needed... so if I had to pick one, it's stick to keeping the allow list (`--only-todo-types`) as that's the one I'm _really_ in need of. Please confirm if I should drop the exclude one.

> The language around packs has moved to calling not calling them violations but rather "todos." This code should adopt that language.

Gotcha. Will change it.

> I also believe these selectors should be grouped with the other todo option, not with the packs exclusion

Sure thing.

[the remark this answers is not in the document]
Comment (shageman) [reply bin/visualize_packs:30]: > so if I had to pick one, it's stick to keeping the allow list (--only-todo-types) as that's the one I'm really in need of

let's go with this one only for now. If we don't have to, I'd rather not introduce the weirdness around being able to include and exclude at the same time

**Comment by shageman** [comment]
Food for (future) thought: It occurs to me that we could limit the legend to the arrows that are actually in the diagram (or at least the ones that are not excluded)

File: lib/visualize_packs.rb (hunk 2 above)
Comment (oboxodo) [reply lib/visualize_packs.rb]: added!

**Comment by oboxodo** [comment]
@shageman I agree about limiting the legend only to displayed types... I did a quick try but I think that can be left for a separate PR. What do you think?

**Review by shageman — APPROVED** [review]

**Pull request merged**